### ISSUE TITLE
Writer chain should own the responsibility of flushing underlying streams

### ISSUE THREAD

**Issue by nesohc** [thymeleaf/thymeleaf#501]
Hi, I'm quite a new user to Thymeleaf so it might be that I am missing something. 

But when testing it together with spring boot I have found that it seems to work "by accident" most times. 
This may very well be wrong but from my understanding the AbstractGeneralTemplateWriter writes to my underlying Writer which I pass on. But if I for example pass on a PrintWriter to it, the PrintWriter will use an underlying Writer, in my case an OutputStreamWriter, which in turn writes to a StreamEncoder.
This StreamEncoder writes to an internal HeapByteBuffer, as an optimization instead of writing every byte it's underlying stream. But that also means that unless it gets flushed the StreamEncoder will never write to the actual underlying stream. 

What I mean with "it works by accident", is that both Jetty and Tomcat is not using a PrintWriter but their own extension of PrintWriter, and as such they never get this chain.

What happens then?
My html never gets printed. 

Why does this happen?
I as a user may want to create my own filter to do something with the stream, in my test application I just create a "cache" of whatever is written, and log that. 

So basically I have 3 solutions to solve this in my application.
1. I can write my own PrintWriter.
2. I can use another PrintWriter extension, like the tomcat (public access) or jetty (private access).
3. I can flush the stream myself, the simplest solution.

The way I would like it to be solved, is that I think the AbstractGeneralTemplateWriter or part of that writing chain should be responsible for flushing the underlying streams to avoid issues like these. 

I have also attached a zip file containing a project with a bunch of TODOs to test and see what I'm trying to explain here. 

[clean-filter.zip](https://github.com/thymeleaf/thymeleaf/files/284475/clean-filter.zip)

Best Regards,
Markus

**Comment by danielfernandez** [comment]
You are right, thanks for pointing this out.

Actually, Thymeleaf 3.0 already performs a `writer.flush()` after processing a template (see https://github.com/thymeleaf/thymeleaf/blob/3.0-master/src/main/java/org/thymeleaf/TemplateEngine.java#L1107-L1114 ) but Spring Boot still uses Thymeleaf 2.1.4, and this does not happen in the 2.1 branch yet.

**Comment by danielfernandez** [comment]
Already in `2.1.5-SNAPSHOT`
